dds/DCPS/XTypes/DynamicTypeMember.cpp

```diff
@@ -0,0 +1,89 @@
+#include "DynamicTypeMember.h"
+#include "MemberDescriptor.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+namespace OpenDDS {
+namespace XTypes {
+
+DynamicTypeMember::DynamicTypeMember()
+: descriptor_(new MemberDescriptor())
+{}
+
+DynamicTypeMember::~DynamicTypeMember()
+{
+  delete descriptor_;
+}
+
+DDS::ReturnCode_t DynamicTypeMember::get_descriptor(MemberDescriptor& descriptor)
+{
+  descriptor = *descriptor_;
+  return DDS::RETCODE_OK;
+}
+
+bool DynamicTypeMember::equals(const DynamicTypeMember& other)
+{
+  DynamicTypePtrPairSeen dt_ptr_pair;
+  return test_equality(*this, other, dt_ptr_pair);
+}
+
+MemberId DynamicTypeMember::get_id()
+{
+  return descriptor_->id;
+}
+
+OPENDDS_STRING DynamicTypeMember::get_name()
+{
+  return descriptor_->name;
+}
+
+bool test_equality(const DynamicTypeMember& lhs, const DynamicTypeMember& rhs, DynamicTypePtrPairSeen& dt_ptr_pair)
+{
+  return test_equality(*lhs.descriptor_, *rhs.descriptor_, dt_ptr_pair);
+}
+
+bool test_equality(const DynamicTypeMembersByName& lhs, const DynamicTypeMembersByName& rhs, DynamicTypePtrPairSeen& dt_ptr_pair)
+{
+  if (lhs.size() == rhs.size()) {
+    if (lhs.size() > 0) {
+      for (DynamicTypeMembersByName::const_iterator it = lhs.begin(); it != lhs.end(); ++it) {
+        if(rhs.find(it->first) != rhs.end()) {
+          if (!(test_equality(*lhs.find(it->first)->second->descriptor_, *rhs.find(it->first)->second->descriptor_, dt_ptr_pair))) {
+            ACE_DEBUG((LM_DEBUG, ACE_TEXT("DynamicTypeMembersByName: Members are not the same\n")));
+            return false;
+          }
+        } else {
+          return false;
+        }
+      }
+    }
+  } else {
+    ACE_DEBUG((LM_DEBUG, ACE_TEXT("DynamicTypeMembersByName: Member count is different\n")));
+    return false;
+  }
+  return true;
```

Comment:
```suggestion
  if (lhs.size() == rhs.size()) {
    for (DynamicTypeMembersByName::const_iterator it = lhs.begin(); it != lhs.end(); ++it) {
      if(rhs.find(it->first) != rhs.end()) {
        if (!test_equality(*lhs.find(it->first)->second->descriptor_, *rhs.find(it->first)->second->descriptor_, dt_ptr_pair)) {
          ACE_DEBUG((LM_DEBUG, ACE_TEXT("DynamicTypeMembersByName: Members are not the same\n")));
          return false;
        }
      } else {
        return false;
      }
    }
    return true;
  } else {
    ACE_DEBUG((LM_DEBUG, ACE_TEXT("DynamicTypeMembersByName: Member count is different\n")));
    return false;
  }
```
Similar changes for the other `test_equality` function.